### PR TITLE
[#1916] Update datastore versioning logic to work with 1.0

### LINKED ISSUE #1916: Update datastore migration code and upgrading.md in preparation for 1.0
Over the last several maintainer syncs, we collectively made the decision that we will break semantic versioning following the release of SPIRE 1.0.0.

The compatibility requirements following 1.0 will match the current compatibility requirements - minor releases function as deprecation boundaries, and support is maintained for n-1 minor release. While this break semantic versioning, it follows suit on precedence set by other large projects such as Kubernetes and Istio.

There is migration code inside the datastore "plugin" that understands version numbers and comptability. Update this code so that it is 1.0-aware, codifying the above policy. As part of this work, we also need to review `upgrading.md` for accuracy.

### PR DESCRIPTION
**Pull request by evan2645**
This commit amends the version compatibility logic in the datastore
migration code to work with 1.0. It applies a targeted exception in
which 0.12.x is declared compatible with 1.0.x as 0.12 will be the last
pre-1.0 release. It additionally moves the major version check to occur
prior to new database initialization so we fail faster in our test suite
when the time for 2.0 comes.

I resisted the urge to refactor the compatibility and auto-migration
safety logic as we are gearing up to release and such a refactor was
decidedly a "nice-to-have" with risk attached. It does need some love
though, perhaps we can clean it up for 0.12.1.

Finally, the issue for this PR also covers the updating of upgrading.md,
but under further review that doc seems more appropriately updated
just prior to the release of 1.0.0. I've created an additional issue to track
that (#2007)

Fixes #1916 

Signed-off-by: Evan Gilman <egilman@vmware.com>

### CODE REVIEW

File: pkg/server/plugin/datastore/sql/migration.go [hunk 1]
```diff
@@ -86,7 +85,7 @@ func migrateDB(db *gorm.DB, dbType string, disableMigration bool, log hclog.Logg
 	}
 
 	if disableMigration {
-		if err = isDisabledMigrationAllowed(dbCodeVersion); err != nil {
+		if err = isDisabledMigrationAllowed(codeVersion, dbCodeVersion); err != nil {
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/migration.go:88]: maybe this is a clue that `codeVersion` should no longer be a package level variable and should be locally scoped to migrateDB.... then all the akwardness around the naming can go away (e.g. `thisCodeVersion`).

**Review by azdagron — APPROVED** [review]
Looks good! Thanks, @evan2645. I have one small nit, but if you want we can address that in the later cleanup you mention.

File: pkg/server/plugin/datastore/sql/migration.go (hunk 1 above)
Comment (evan2645) [reply pkg/server/plugin/datastore/sql/migration.go:88]: 💯 

Yea, this is one of many smells in the compatibility checks that we have in this package.

How about a compromise - I'll stage changes for cleaning all this mess up, and we can merge them immediately following 0.12.0 release?

Comment (azdagron) [reply pkg/server/plugin/datastore/sql/migration.go:88]: Sounds good to me!

**Review by amartinezfayo — APPROVED** [review]

**Pull request merged**